File: py/shinycomponent/_dataframe.py
from __future__ import annotations

import json
from pathlib import PurePath
from typing import TYPE_CHECKING, Optional

import shiny.experimental as x
from htmltools import HTMLDependency, Tag, tags
from shiny._namespaces import resolve_id

from . import __version__
from ._render_dataframe import data_frame

if TYPE_CHECKING:
    import pandas as pd


def output_data_frame(
    id: Optional[str] = None,
    *,
    data: Optional[pd.DataFrame] = None,
) -> Tag:
    """
    Create a output container for a data frame.

    Parameters
    ----------
    id
        An input id.

    data
        An optional Pandas data frame.

    Returns
    -------
    :
        A UI element.
    """

    res = x.ui.as_fillable_container(
        x.ui.as_fill_item(
            Tag(
                "shiny-data-frame",
                dataframe_dep(),
                id=resolve_id(id) if id is not None else None,
            ),
        )
    )

    if data is not None:

        @data_frame
        def data_fn() -> pd.DataFrame:
            return data

        res.children.append(
            tags.script(json.dumps(data_fn()), type="application/json", class_="data"),
        )

    return res


www_path = PurePath(__file__).parent / "www"


def dataframe_dep() -> HTMLDependency:
    return HTMLDependency(
        name="data-frame",
        version=__version__,
        source={
            "package": "shinycomponent",
            "subdir": str(www_path),
        },
        script=[
            {"src": "dataframe.js", "type": "module"},
        ],
    )

File: r/R/dataframe.R
#' @export
output_data_frame <- function(id = NULL, ..., data = NULL) {
  data_tag <- NULL
  if (!is.null(data)) {
    data_tag <- tags$script(
      HTML(toJSON(to_data_frame_format(data, ...), auto_unbox = TRUE)),
      type="application/json",
      class="data"
    )
  }

  tag("shiny-data-frame", list(id = id, dataframe_dep(), data_tag, ...))
}


#' @importFrom shiny installExprFunction createRenderFunction
#' @importFrom jsonlite toJSON
#' @export
render_data_frame <- function(
    expr,
  width = NULL,
  height = NULL,
  row_selection = FALSE,
  column_selection = FALSE,
  cell_selection = FALSE,
  range_selection = FALSE,
  ...,
  env = parent.frame(),
  quoted = FALSE,
  output_args=list()
) {
  func <- installExprFunction(expr, "func", env, quoted, label = "render_data_frame")

  createRenderFunction(
    func,
    function(value, session, name, ...) {
      to_data_frame_format(
        value,
        width = width,
        height = height,
        row_selection = row_selection,
        column_selection = column_selection,
        cell_selection = cell_selection,
        range_selection = range_selection
      )
    },
    output_data_frame,
    output_args
  )
}


to_data_frame_format <- function(
    df,
  ...,
  width = NULL,
  height = "500px",
  style = c("grid", "table"),
  summary = TRUE,
  row_selection_mode = c("none", "single", "multiple")
) {
  row_selection_mode <- match.arg(row_selection_mode)
  style <- match.arg(style)
  nrows <- nrow(df)

  # Convert factor cols to strings. The data frame also becomes a list here.
  df <- lapply(df, function(col) {
    if (is.factor(col)) as.character(col)
    else col
  })

  res <- list(
    columns = names(df),
    index = seq_len(nrows),
    data = .mapply(list, unname(df), NULL),
    options = list(
      width = width,
      height = height,
      row_selection_mode = row_selection_mode,
      style = style
    )
  )

  res
}


dataframe_dep <- function() {
  htmlDependency(
    name = "data-frame",
    version = as.character(packageVersion("shinycomponent")),
    src = system.file(package = "shinycomponent", "www"),
    script = list(src = "dataframe.js", type = "module")
  )
}
